### DISCUSSION TITLE
s2: corrupt input  - compression

### DISCUSSION THREAD

**Discussion by AlexZIX** [minio/minio#14393] [Q&A]
Today I got stuck because can't download few files and get "corrupt error" in logs of my distributed minio. Monitoring shows me that all nodes and disks are healty. I've tried to download it using minio web console and using geesefs fuse filesystem. Both tries was unsuccessful.

## Expected Behavior
Can download any file from my minio.

## Current Behavior
Can't download few files and got "corrupt error" in logs.

Feb 23 01:15:38 minio-1 minio[178430]: API: GetObject(bucket=wp3d-angstrem-archive, object=webplanner/10/48/1048242.b3db)
Feb 23 01:15:38 minio-1 minio[178430]: Time: 22:15:38 UTC 02/22/2022
Feb 23 01:15:38 minio-1 minio[178430]: DeploymentID: fca050d1-90fc-4b6c-898c-ff98f7f067a6
Feb 23 01:15:38 minio-1 minio[178430]: RequestID: 16D63B80B81AF7C4
Feb 23 01:15:38 minio-1 minio[178430]: RemoteHost: 192.168.155.247
Feb 23 01:15:38 minio-1 minio[178430]: Host: minio-1.bazis.net:9001
**Feb 23 01:15:38 minio-1 minio[178430]: UserAgent: GeeseFS/0.30.4 (go1.16.10; linux; amd64)**
Feb 23 01:15:38 minio-1 minio[178430]: Error: Unable to write all the data to client s2: corrupt input (*fmt.wrapError)
Feb 23 01:15:38 minio-1 minio[178430]:        3: cmd/object-handlers.go:522:cmd.objectAPIHandlers.getObjectHandler()
Feb 23 01:15:38 minio-1 minio[178430]:        2: cmd/object-handlers.go:579:cmd.objectAPIHandlers.GetObjectHandler()
Feb 23 01:15:38 minio-1 minio[178430]:        1: net/http/server.go:2047:http.HandlerFunc.ServeHTTP()

Feb 23 01:17:00 minio-1 minio[178430]: API: GetObject(bucket=wp3d-angstrem-archive, object=webplanner/10/48/1048242.b3db)
Feb 23 01:17:00 minio-1 minio[178430]: Time: 22:17:00 UTC 02/22/2022
Feb 23 01:17:00 minio-1 minio[178430]: DeploymentID: fca050d1-90fc-4b6c-898c-ff98f7f067a6
Feb 23 01:17:00 minio-1 minio[178430]: RequestID: 16D63B93CB752FD3
Feb 23 01:17:00 minio-1 minio[178430]: RemoteHost: 192.168.155.248
Feb 23 01:17:00 minio-1 minio[178430]: Host: minio-1.bazis.net:9001
**Feb 23 01:17:00 minio-1 minio[178430]: UserAgent: MinIO (linux; amd64) minio-go/v7.0.22**
Feb 23 01:17:00 minio-1 minio[178430]: Error: s2: corrupt input (*errors.errorString)
Feb 23 01:17:00 minio-1 minio[178430]:        5: cmd/api-errors.go:2112:cmd.toAPIErrorCode()
Feb 23 01:17:00 minio-1 minio[178430]:        4: cmd/api-errors.go:2137:cmd.toAPIError()
Feb 23 01:17:00 minio-1 minio[178430]:        3: cmd/object-handlers.go:518:cmd.objectAPIHandlers.getObjectHandler()
Feb 23 01:17:00 minio-1 minio[178430]:        2: cmd/object-handlers.go:579:cmd.objectAPIHandlers.GetObjectHandler()
Feb 23 01:17:00 minio-1 minio[178430]:        1: net/http/server.go:2047:http.HandlerFunc.ServeHTTP()

## Steps to Reproduce (for bugs)
I don't know how to reproduce this because I can download some files but at the same time can't download another files in the same bucket.

## Regression
Found this problem when upgrade my minio to 2022-02-08 from 2022-11-02.

## Your Environment
* Version used (`minio --version`): 2022-02-18T01:50:10Z
* Server setup and configuration: 3 nodes with 4 disks each
* Operating System and version (`uname -a`): Linux minio-1 5.4.0-91-generic #102-Ubuntu SMP Fri Nov 5 16:31:28 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by harshavardhana** [comment]
> Feb 23 01:15:38 minio-1 minio[178430]: API: GetObject(bucket=wp3d-angstrem-archive, object=webplanner/10/48/1048242.b3db) Feb 23 01:15:38 minio-1 minio[178430]: Time: 22:15:38 UTC 02/22/2022 Feb 23 01:15:38 minio-1 minio[178430]: DeploymentID: fca050d1-90fc-4b6c-898c-ff98f7f067a6 Feb 23 01:15:38 minio-1 minio[178430]: RequestID: 16D63B80B81AF7C4 Feb 23 01:15:38 minio-1 minio[178430]: RemoteHost: 192.168.155.247 Feb 23 01:15:38 minio-1 minio[178430]: Host: minio-1.bazis.net:9001 **Feb 23 01:15:38 minio-1 minio[178430]: UserAgent: GeeseFS/0.30.4 (go1.16.10; linux; amd64)** Feb 23 01:15:38 minio-1 minio[178430]: Error: Unable to write all the data to client s2: corrupt input (*fmt.wrapError) Feb 23 01:15:38 minio-1 minio[178430]: 3: cmd/object-handlers.go:522:cmd.objectAPIHandlers.getObjectHandler() Feb 23 01:15:38 minio-1 minio[178430]: 2: cmd/object-handlers.go:579:cmd.objectAPIHandlers.GetObjectHandler() Feb 23 01:15:38 minio-1 minio[178430]: 1: net/http/server.go:2047:http.HandlerFunc.ServeHTTP()
> 
> Feb 23 01:17:00 minio-1 minio[178430]: API: GetObject(bucket=wp3d-angstrem-archive, object=webplanner/10/48/1048242.b3db) Feb 23 01:17:00 minio-1 minio[178430]: Time: 22:17:00 UTC 02/22/2022 Feb 23 01:17:00 minio-1 minio[178430]: DeploymentID: fca050d1-90fc-4b6c-898c-ff98f7f067a6 Feb 23 01:17:00 minio-1 minio[178430]: RequestID: 16D63B93CB752FD3 Feb 23 01:17:00 minio-1 minio[178430]: RemoteHost: 192.168.155.248 Feb 23 01:17:00 minio-1 minio[178430]: Host: minio-1.bazis.net:9001 **Feb 23 01:17:00 minio-1 minio[178430]: UserAgent: MinIO (linux; amd64) minio-go/v7.0.22** Feb 23 01:17:00 minio-1 minio[178430]: Error: s2: corrupt input (*errors.errorString) Feb 23 01:17:00 minio-1 minio[178430]: 5: cmd/api-errors.go:2112:cmd.toAPIErrorCode() Feb 23 01:17:00 minio-1 minio[178430]: 4: cmd/api-errors.go:2137:cmd.toAPIError() Feb 23 01:17:00 minio-1 minio[178430]: 3: cmd/object-handlers.go:518:cmd.objectAPIHandlers.getObjectHandler() Feb 23 01:17:00 minio-1 minio[178430]: 2: cmd/object-handlers.go:579:cmd.objectAPIHandlers.GetObjectHandler() Feb 23 01:17:00 minio-1 minio[178430]: 1: net/http/server.go:2047:http.HandlerFunc.ServeHTTP()

This error is generated due to an incoming stream input for S2 decompression that is failing.  Can you provide all the server environment variables that are set? and also `mc admin inspect alias/wp3d-angstrem-archive/webplanner/10/48/1048242.b3db` ?

**Comment by AlexZIX** [comment]
This is my ENVs for minio:

MINIO_ROOT_USER="ZiX"
MINIO_ROOT_PASSWORD="*******"
MINIO_OPTS="-C /etc/minio --address :9001 --console-address :8081"
MINIO_VOLUMES="http://minio-{1...3}.bazis.net/disks/hdd{6...9}"
MINIO_REGION_NAME="ru-central"

Also compression enabled for .b3db files in minio web console.

You can download inspect file here: https://disk.yandex.ru/d/CUTVWPc_pUb80Q

**Comment by AlexZIX** [comment]
Found this in log. May be it will help.

Feb 23 01:43:31 minio-1 minio[290355]: API: GetObject(bucket=wp3d-angstrem-archive, object=webplanner/10/48/1048088.b3db)
Feb 23 01:43:31 minio-1 minio[290355]: Time: 22:43:31 UTC 02/22/2022
Feb 23 01:43:31 minio-1 minio[290355]: DeploymentID: fca050d1-90fc-4b6c-898c-ff98f7f067a6
Feb 23 01:43:31 minio-1 minio[290355]: RequestID: 16D63D065A76B220
Feb 23 01:43:31 minio-1 minio[290355]: RemoteHost: 192.168.155.248
Feb 23 01:43:31 minio-1 minio[290355]: Host: minio-1.bazis.net:9001
Feb 23 01:43:31 minio-1 minio[290355]: UserAgent: MinIO (linux; amd64) minio-go/v7.0.22
Feb 23 01:43:31 minio-1 minio[290355]: Error: s2: corrupt input, **crc mismatch** (*errors.errorString)
Feb 23 01:43:31 minio-1 minio[290355]:        5: cmd/api-errors.go:2112:cmd.toAPIErrorCode()
Feb 23 01:43:31 minio-1 minio[290355]:        4: cmd/api-errors.go:2137:cmd.toAPIError()
Feb 23 01:43:31 minio-1 minio[290355]:        3: cmd/object-handlers.go:518:cmd.objectAPIHandlers.getObjectHandler()
Feb 23 01:43:31 minio-1 minio[290355]:        2: cmd/object-handlers.go:579:cmd.objectAPIHandlers.GetObjectHandler()
Feb 23 01:43:31 minio-1 minio[290355]:        1: net/http/server.go:2047:http.HandlerFunc.ServeHTTP()

**Comment by harshavardhana** [comment]
@AlexZIX please capture `mc admin config get alias/ compression` values as well.

**Comment by harshavardhana** [comment]
Can you get inspect for `mc admin inspect alias/wp3d-angstrem-archive/webplanner/10/48/1048242.b3db/**`

**Comment by AlexZIX** [comment]
> mc admin config get alias/ compression

@harshavardhana this is output:
compression allow_encryption=off extensions=.txt,.log,.csv,.json,.tar,.xml,.bin,.b3d,.b3db,.ldw mime_types=text/*,application/json,application/xml,binary/octet-stream

This is a new inspect archive: https://disk.yandex.ru/d/t9XYzFkn60RqyA

**Comment by klauspost** [comment]
I think this is related to healing rather than compression. 4 of the shards shows they were healed (and at least 2 disks were missing during write). 

It just shows the error rather than return the bad data. The bad crc error also matches that the corruption is right at the end, since that is where the CRC is placed.

@harshavardhana Did we make a tool to correct these healing errors, or should the shards just be deleted and re-healed with a new version?

**Comment by AlexZIX** [comment]
@klauspost Does that means that I lost this files? Or I should wait until healing will finish or new version with fix for such situation will be released?
You said that 2 disks were missing during write but after replacing 4 disks at one node I haven't got any errors. This files was created before I've replaced disks. I thought that minio should return error when writing of file are impossible because disks missing and instance haven't write quorum. Is that right or I misunderstand something?

**Comment by klauspost** [comment]
@AlexZIX I cannot tell for sure. It should be recoverable, though maybe not automatically. Harsha could have input on this.

**Comment by harshavardhana** [comment]
> @harshavardhana Did we make a tool to correct these healing errors, or should the shards just be deleted and re-healed with a new version?

@klauspost they will only re-heal with the disk timestamp difference - this is perhaps related to the padding issue? it may not be possible in all cases.

**Comment by harshavardhana** [comment]
The way manually it can be fixed is by removing the tail end of the shard i.e figure out which is the last data shard and then reheal it.

**Comment by harshavardhana** [comment]
I will paste here how to figure this out @AlexZIX so it might be helpful for you later today.

**Comment by harshavardhana** [comment]
these files were possibly healed incorrectly before `RELEASE.2021-12-09T06-19-41Z`  this release.

**Reply by harshavardhana** [reply]
> these files were possibly healed incorrectly before `RELEASE.2021-12-09T06-19-41Z` this release.

Actually, these objects were created right before this release - mainly exposing themselves to the potential padding issue.

**Reply by AlexZIX** [reply]
@harshavardhana Healing at my minio was started at January 2022 and finished at February. After that according to your advice I've upgraded minio to the latest version.
Does that means that all users whos instances doing healing before RELEASE.2021-12-09T06-19-41Z now have corrupted files?

**Reply by harshavardhana** [reply]
> Does that means that all users whos instances doing healing before RELEASE.2021-12-09T06-19-41Z now have corrupted files?

It depends on what was healed, it doesn't affect all objects only the ones where the last data shard drive was replaced. Also simply removing those parts will heal them as well correctly.

**Reply by AlexZIX** [reply]
> It depends on what was healed, it doesn't affect all objects only the ones where the last data shard drive was replaced. Also simply removing those parts will heal them as well correctly.

Will you teach me how to do that? Just erase all drives which was healed and let minio heal it again?

**Comment by harshavardhana** [comment]
can you also capture `mc admin inspect alias/.minio.sys/format.json`? @AlexZIX

**Reply by AlexZIX** [reply]
@harshavardhana You can download inspect file here: https://disk.yandex.ru/d/EtXj6hf0dxHLUQ

**Comment by harshavardhana** [comment]
can you share your server command line as well @AlexZIX ?

**Reply by AlexZIX** [reply]
> can you share your server command line as well @AlexZIX ?

@harshavardhana Sent login details to your e-mail harsha@minio.io

**Reply by harshavardhana** [reply]
> @harshavardhana Sent login details to your e-mail [harsha@minio.io](mailto:harsha@minio.io)

@AlexZIX I can't log in to your clusters without having commercial contract with us.

**Reply by AlexZIX** [reply]
> command

@harshavardhana  Seems I understand you wrong. This is command line which I use to start minio.

/usr/local/bin/minio server -C /etc/minio --address :9001 --console-address :8081 http://minio-{1...3}.bazis.net/disks/hdd{6...9}

**Reply by AlexZIX** [reply]
@harshavardhana So is solution exists for now or I should wait for the next release where it may be fixed?

**Reply by harshavardhana** [reply]
> @harshavardhana So is solution exists for now or I should wait for the next release where it may be fixed?

I tried unfortunately your data is corrupted beyond repair, I can't see how it can be recovered. 

We have automation in place for this but your data doesn't satisfy those constraints of recovery.

**Reply by AlexZIX** [reply]
May be I can try to erase one disk and let minio heal it? Or data broken at parity disks too?

**Reply by harshavardhana** [reply]
> May be I can try to erase one disk and let minio heal it? Or data broken at parity disks too?

It is not predictable, there is no easier way to know - it requires a lot of time. I just don't have the time right now.

**Comment by AlexZIX** [comment]
I see that minio started healing again. I haven't changed anything in my configuration. May this process recover files which unavailable for downloading now?

root@minio-1:~# mc admin heal minio-cold
Objects Healed: 293,149/874,739 (33.5%%), 4.3 TB/12 TB (34.9%%)
Heal rate: 0 obj/s, 6.9 KiB/s
Estimated Completion: 2 months
